### ISSUE TITLE
Black screen with amdgpu

### ISSUE THREAD

**Issue by ivan-m** [NixOS/nixpkgs#54050]
## Issue description

I last updated my nixos system (new system, Ryzen 2700X, Gigabyte Aurus X470 motherboard, Radeon Vega 64 GPU) in October 2018.  At that point in time booting worked perfectly.

When I now do a `nix-channel --update` and `nixos-rebuild switch` then reboot, after some boot messages I'm left with a black screen with a non-blinking cursor in the top left corner.  Quite often not even SSH will work.

(Rolling back to the channel I had in October and rebuilding fixes it again.  Using the corresponding nixos-unstable kernel version from then works as well.)

I've tried changing kernel version (including what's in nixos-unstable), the solutions in #44813, etc. to no avail.

Full journalctl logs from a failing boot can be found [here](https://gist.github.com/ivan-m/89faa89e2c45219ccc2a7aea6cec52ea).

### Steps to reproduce

Update channels and system.

## Technical details

Working system:

```bash
# nix-shell -p nix-info --run "nix-info -m"
warning: Nix search path entry '$HOME/.nix-defexpr/channels' does not exist, ignoring
these paths will be fetched (0.05 MiB download, 0.28 MiB unpacked):
  /nix/store/kic17fw8wil74k04kzh3dha43izrr9j3-bash-interactive-4.4-p23-dev
copying path '/nix/store/kic17fw8wil74k04kzh3dha43izrr9j3-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
warning: Nix search path entry '$HOME/.nix-defexpr/channels' does not exist, ignoring
 - system: `"x86_64-linux"`
 - host os: `Linux 4.18.14, NixOS, 18.09beta78.6f3bd5db2f4 (Jellyfish)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.0.4`
 - channels(root): `"nixos-18.09beta78.6f3bd5db2f4, unstable-19.03pre155751.45a419ab5a2"`
 - channels(ivan): `""`
warning: Nix search path entry '$HOME/.nix-defexpr/channels' does not exist, ignoring
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

After updating (but not rebooting):

```bash
# nix-shell -p nix-info --run "nix-info -m"
warning: Nix search path entry '$HOME/.nix-defexpr/channels' does not exist, ignoring
these paths will be fetched (0.05 MiB download, 0.28 MiB unpacked):
  /nix/store/4m253khy42mqrv3l3kih6ar75fvi0wl5-bash-interactive-4.4-p23-dev
copying path '/nix/store/4m253khy42mqrv3l3kih6ar75fvi0wl5-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
warning: Nix search path entry '$HOME/.nix-defexpr/channels' does not exist, ignoring
 - system: `"x86_64-linux"`
 - host os: `Linux 4.18.14, NixOS, 18.09.1922.97e0d53d669 (Jellyfish)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.3`
 - channels(root): `"nixos-18.09.1922.97e0d53d669, unstable-19.03pre166308.626233eee6e"`
 - channels(ivan): `""`
warning: Nix search path entry '$HOME/.nix-defexpr/channels' does not exist, ignoring
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by clhodapp** [comment]
I have the same symptoms with the same GPU. I'm using a similar mitigation: booting from an old system profile and checking from time to time to see if the issue has ironed itself out upstream. Personally, I'm tracking the 4.19 kernel in nixpkgs for nixos 18.09.

**Comment by clhodapp** [comment]
My /boot just filled up in my latest attempt to check the state of master, which prompted me to have to clean out old profiles. That motivated me to bother to track the nixpkgs github repo for my kernels so I could make myself a fallback of a known-working kernel (4.19.5). Since I went to the trouble to do that, I also tried the latest 4.20 kernel from nixpkgs stable (currently at 4.20.5, since 4.20.6 hasn't made it to the channel yet).

Long story short: it works! So perhaps time will fix the issue for you too and perhaps kernel 4.20 should be brought into the 18.09 branch of nixpkgs.

I suspect that one of the many powerplay fixes that AMD brought in for 4.20 fixed things but I'm not willing to do the deep analysis needed to confirm. Overall, my experience is that AMD support in the kernel has been pretty buggy for me!

**Comment by ivan-m** [comment]
@clhodapp thank you so much for figuring this out!

For the record, my implementation of this is:

```nix
let
  unstable = (import <unstable> {
    config = {allowUnfree = true;};
  });
in
{
  nixpkgs.config = {
    packageOverrides = super: let self = super.pkgs; in {
      linuxPackages_latest = unstable.linuxPackages_latest;
    };
  };
}
```

And then using `boot.kernelPackages = pkgs.linuxPackages_latest;`.

I too would appreciate 4.20 being brought into 18.09.

**Comment by clhodapp** [comment]
Is it possible that you overrode `linuxPackages_latest` for your user rather than for the system?

**Comment by clhodapp** [comment]
Oh nice! That's great! I misunderstood and thought you were giving more detail on "I've tried changing kernel version (including what's in nixos-unstable)" from the initial issue. Glad you got your system working!

**Comment by arianvp** [comment]
this seems fixed on `20.03` . Vega support should be fine. Just tried this with a Radeon VII and it worked.

**Comment by davidak** [comment]
I successfully use a Radeon RX 480 with amdgpu driver on 20.03. Can not reproduce the problem, only when still use nvidia driver.

**Comment by spiderbit** [comment]
@arianvp I just use here Ryzen 2400g and first I get black screen till I edit grub line to add nomodeline.

Then X starts stubbern in vesa mode no matter what I do using 20.03.

```
  nixpkgs.config.allowUnfree = true;
  boot.kernelPackages = pkgs.linuxPackages_latest;
  services.xserver.videoDrivers = ["amdgpu" "radeon" "cirrus" "vesa" "vmware" "modesetting" ];
  hardware.cpu.amd.updateMicrocode = true;
  environment.systemPackages = with pkgs; [
...microcodeAmd radeontop xorg.xf86videoamdgpu
];
```

**Comment by spiderbit** [comment]
ok I don't know how much of the lines are nessesary but what fixed it for me after trying like 20 kernel options was to replace the displayport cable with a hdmi cable...

It's strange if it's defect or not 1.4 compatible you would expect it not to work at all, not that it runs with mesa just fine more or less with full resolution... Maybe some displayport specific firmware things... I don't know but hdmi works without any kernelparams.

```
  nixpkgs.config.allowUnfree = true;
  services.xserver.modules = [pkgs.xorg.xf86videoamdgpu];
  hardware.enableRedistributableFirmware = true;
```

This are the last 3 things I have in my config now didn't test it without them, maybe do test later if any of them are needed.

**Comment by samueldr** [comment]
Given this is an issue around fresh new hardware *at the time*, and it's been over three years since the last update, I'll tentatively close this as resolved. Anyway this would have been a kernel-side thing to fix, and not specific to NixOS.

To the people originally involved: if any of you is still having this precise issue with the hardware listed previously, please say so, I will re-open.

To anyone else having similar issues: First, try the workarounds/solutions listed here (using the latest kernel package, using `nomodeset`). Then, please open a fresh new issue, detailing your particular circumstances. Feel free to cross-reference this issue at the same time.
